### PR TITLE
NFI Pallet 

### PR DESCRIPTION
**Pull request by JasonTulp**
Adds the NFI pallet. This pallet is responsible for storing the mappings between NFT and SFT tokenIds and an IPFS link which points to the AI related data such as murmur matrix and Genome matrix

https://futureverse.atlassian.net/browse/TRN-466?atlOrigin=eyJpIjoiNWY4OGE5ZGJiNzMwNDY5ZDlkZDVkYTg0MzA5ZjM2YTYiLCJwIjoiaiJ9

### CODE REVIEW

File: pallet/common/src/test_utils.rs
Comment (zees-dev) [inline pallet/common/src/test_utils.rs:1]: nice impl for unit type

File: pallet/nfi/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+pub use pallet::*;
+
+mod types;
+use types::*;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+pub mod mock;
```
Comment (zees-dev) [inline pallet/nfi/src/lib.rs:37]: nit: why exposed to pub api
```suggestion
mod mock;
```

File: pallet/nfi/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+	/// The permission enabled relayer
+	#[pallet::storage]
+	pub type Relayer<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
+
+	/// The extra cost to cover
+	#[pallet::storage]
+	pub type MintFee<T: Config> = StorageMap<_, Twox64Concat, NFISubType, FeeDetails<T::AccountId>>;
+
+	#[pallet::storage]
+	pub type NfiData<T: Config> = StorageDoubleMap<
```
Comment (zees-dev) [inline pallet/nfi/src/lib.rs:83]: nit: acronyms should be capitalised imo (i.e. `Nfi` -> `NFI`) - similar to what we have for the nft pallet. But seems like we don't do the same in sft :/

File: pallet/nfi/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+		_,
+		Twox64Concat,
+		TokenId,
+		Twox64Concat,
+		NFISubType,
+		NFIDataType<T::MaxDataLength>,
+	>;
+
+	#[pallet::storage]
+	pub type NfiEnabled<T> = StorageDoubleMap<
```
Comment (zees-dev) [inline pallet/nfi/src/lib.rs:93]: can this be a single storage map? I.e. if the item exists here, then its enabled, if not, then it's not enabled?

File: pallet/nfi/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+		#[pallet::call_index(4)]
+		#[pallet::weight(T::WeightInfo::manual_data_request())]
+		pub fn manual_data_request(
+			origin: OriginFor<T>,
+			token_id: TokenId,
+			sub_type: NFISubType,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(NfiEnabled::<T>::get(token_id.0, sub_type), Error::<T>::NotEnabled);
+			// TODO maybe go for token owner or collection owner. Leni to confirm
```
Comment (zees-dev) [inline pallet/nfi/src/lib.rs:234]: todo needs to be addressed

File: pallet/nfi/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+			caller,
+			sub_type,
+			collection_id,
+			serial_numbers,
+		});
+	}
+
+	/// Returns true if who is the owner of the collection.
+	/// Checks both NFT and SFT pallet
+	fn is_collection_owner(collection_id: CollectionUuid, who: &T::AccountId) -> bool {
```
Comment (zees-dev) [inline pallet/nfi/src/lib.rs:327]: note: this implies `CollectionUuid` is unique across both pallets

File: pallet/nft/src/lib.rs [hunk 6]
```diff
@@ -49,6 +52,8 @@ use sp_std::prelude::*;
 mod benchmarking;
 #[cfg(test)]
 pub mod mock;
+#[cfg(feature = "std")]
```
Comment (zees-dev) [inline pallet/nft/src/lib.rs:55]: only import for tests instead?
```suggestion
#[cfg(test)]
```

File: pallet/sft/src/lib.rs [hunk 7]
```diff
@@ -28,6 +30,8 @@ use sp_std::prelude::*;
 
 #[cfg(test)]
 pub mod mock;
+#[cfg(feature = "std")]
```
Comment (zees-dev) [inline pallet/sft/src/lib.rs:33]: same here

File: pallet/sft/src/mock.rs [hunk 8]
```diff
@@ -35,6 +35,12 @@ impl_pallet_assets_config!(Test);
 impl_pallet_assets_ext_config!(Test);
 impl_pallet_nft_config!(Test);
 
+pub struct MockTransferSubscriber;
+
+impl OnTransferSubscriber for MockTransferSubscriber {
+	fn on_nft_transfer(_token_id: &TokenId) {}
+}
```
Comment (zees-dev) [inline pallet/sft/src/mock.rs:42]: why not use unit `()` type instead?

File: pallet/nfi/src/lib.rs (hunk 5 above)
Comment (JasonTulp) [reply pallet/nfi/src/lib.rs:327]: Yep, CollectionUuid is guaranteed to be unique across NFT and SFT pallets

File: pallet/nft/src/lib.rs (hunk 6 above)
Comment (JasonTulp) [reply pallet/nft/src/lib.rs:55]: I think we talked about this previously but other pallets can't access the test_utils if it is behind the test flag, hence why we decided to use std

File: pallet/nfi/src/lib.rs (hunk 4 above)
Comment (JasonTulp) [reply pallet/nfi/src/lib.rs:234]: Good catch, waiting for response

File: pallet/nfi/src/lib.rs (hunk 3 above)
Comment (JasonTulp) [reply pallet/nfi/src/lib.rs:93]: Not really, the NfiEnabled has collection id as a key and denotes whether Nfi is enabled for the entire collection, however NfiData is per token so they kind of need to be separated

File: pallet/nfi/src/lib.rs (hunk 1 above)
Comment (JasonTulp) [reply pallet/nfi/src/lib.rs]: Whoops, this was left over after some experiments. Addressed: https://github.com/futureversecom/trn-seed/pull/866/commits/e5e9de4910537d3d4c8f71c18510063a0d9d3f78

File: pallet/sft/src/mock.rs (hunk 8 above)
Comment (JasonTulp) [reply pallet/sft/src/mock.rs]: https://github.com/futureversecom/trn-seed/pull/866/commits/e5e9de4910537d3d4c8f71c18510063a0d9d3f78

File: pallet/nfi/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+description = "Root Network NFI pallet"
+
+[dependencies]
+seed-primitives = { path = "../../primitives", default-features = false }
+seed-pallet-common = { path = "../common", default-features = false }
+
+codec = { package = "parity-scale-codec", version = "3.0.0", default-features = false, features = ["derive"] }
+scale-info = { version = "2.3.0", default-features = false, features = ["derive"] }
+
+sp-core = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.30" }
```
Comment (surangap) [inline pallet/nfi/Cargo.toml:17]: update as per new main. use the workspace as required

File: pallet/nfi/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+	// For SFT it only checks whether who is the owner of the collection
+	// This is due to SFT tokens being owned by the collection owner, where users can have some
+	// balance of the token
+	fn check_permissions(token_id: TokenId, who: &T::AccountId) -> bool {
+		if let Some(nft_owner) = T::NFTExt::get_token_owner(&token_id) {
+			if who == &nft_owner {
+				return true;
+			}
+			// Not token owner, check if who is the collection owner
+			if let Ok(nft_owner) = T::NFTExt::get_collection_owner(token_id.0) {
```
Comment (surangap) [inline pallet/nfi/src/lib.rs:347]:
```suggestion
			if let Ok(collection_owner) = T::NFTExt::get_collection_owner(token_id.0) {
```

File: pallet/nfi/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+		#[pallet::weight(T::WeightInfo::manual_data_request())]
+		pub fn manual_data_request(
+			origin: OriginFor<T>,
+			token_id: TokenId,
+			sub_type: NFISubType,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(NfiEnabled::<T>::get(token_id.0, sub_type), Error::<T>::NotEnabled);
+			// Check that the caller is the token or collection owner
+			ensure!(Self::check_permissions(token_id.clone(), &who), Error::<T>::NotTokenOwner);
```
Comment (surangap) [inline pallet/nfi/src/lib.rs:235]: `check_permissions()` returns true for either token owner or collection owner. The description says
``` /// Caller must be the owner of the token```

Comment (surangap) [reply pallet/nfi/src/lib.rs:235]: as per the test https://github.com/futureversecom/trn-seed/pull/866/files#diff-e61e708f2babdfa6d60bfa084fa92437a59b96fcfc4b15a5dbb057c273d52160R262, it seems either should be able to call it. in that case update the doc comment?

**Review by surangap — COMMENTED** [review]
overall lgtm, nice work. Added a couple of comments.

File: pallet/nfi/src/lib.rs (hunk 4 above)
Comment (JasonTulp) [reply pallet/nfi/src/lib.rs]: Updated to allow for both

Comment (JasonTulp) [reply pallet/nfi/src/lib.rs]: https://github.com/futureversecom/trn-seed/pull/866/commits/979c6ac7174f2a3cb7f412d0f589869ea35c0f96

File: pallet/nfi/src/lib.rs (hunk 11 above)
Comment (JasonTulp) [reply pallet/nfi/src/lib.rs:237]: Good catch, this was added later and I missed the comment of this extrinsic
https://github.com/futureversecom/trn-seed/pull/866/commits/7e8b377ea67d1bbadc28a84cabce2aa0b6c62cde

File: pallet/nfi/src/lib.rs (hunk 10 above)
Comment (JasonTulp) [reply pallet/nfi/src/lib.rs]: https://github.com/futureversecom/trn-seed/pull/866/commits/7e8b377ea67d1bbadc28a84cabce2aa0b6c62cde

File: pallet/nfi/Cargo.toml (hunk 9 above)
Comment (JasonTulp) [reply pallet/nfi/Cargo.toml]: https://github.com/futureversecom/trn-seed/pull/866/commits/f0b899c63991838c30b0f5361b5abecffc2111fe

**Review by surangap — APPROVED** [review]
Do a CI benchmark run as well.

**Review by zees-dev — APPROVED** [review]
LGTM 👍

**Pull request merged**